### PR TITLE
replace PAUSE / RELEASE with STOP / START

### PR DESCRIPTION
**Pull request by lyakh**
Currently many components, e.g. DAIs handle STOP and PAUSE differently. Also the mixer component blocks PAUSE while obviously accepting and propagating STOP. This creates a lot of complexity in forking pipelines. With START, STOP, PAUSE and RELEASE now handled in the pipeline task context, it should be possible to handle PAUSE and RELEASE exactly like STOP and START.

### CODE REVIEW

File: src/include/sof/audio/component.h [hunk 1]
```diff
@@ -78,20 +78,20 @@ struct timestamp_data;
...
-#define COMP_TRIGGER_PREPARE		7	/**< Prepare component */
-#define COMP_TRIGGER_XRUN		8	/**< XRUN component */
-#define COMP_TRIGGER_PRE_START		9	/**< Prepare to start component stream */
-#define COMP_TRIGGER_PRE_RELEASE	10	/**< Prepare to release paused component stream */
-#define COMP_TRIGGER_POST_STOP		11	/**< Finalize stop component stream */
-#define COMP_TRIGGER_POST_PAUSE		12	/**< Finalize pause component stream */
-#define COMP_TRIGGER_NO_ACTION		13	/**< No action required */
+enum {
+	COMP_TRIGGER_STOP,		/**< Stop component stream */
+	COMP_TRIGGER_START,		/**< Start component stream */
```
Comment (dbaluta) [inline src/include/sof/audio/component.h:83]: I would prefer that you convert this to an enum in another patch. This patch is just about removing the two unused commands.

**Comment by lyakh** [comment]
**FIXME:** the last commit breaks several things:
1. it adds an unbalanced `notifier_unregister()`
2. it probably breaks (again) the now perfectly correct HDA DMA stop flow

All of those must be fixed before merging this

File: src/drivers/imx/esai.c [hunk 2]
```diff
@@ -374,10 +374,6 @@ static int esai_trigger(struct dai *dai, int cmd, int direction)
 	case COMP_TRIGGER_PAUSE:
 		esai_stop(dai, direction);
 		break;
-	/* Remaining triggers are no-ops */
-	case COMP_TRIGGER_SUSPEND:
```
Comment (lgirdwood) [inline src/drivers/imx/esai.c:378]: I assume the suspend removal is in the other PR now ?

File: src/drivers/imx/sai.c [hunk 3]
```diff
@@ -332,8 +332,6 @@ static int sai_trigger(struct dai *dai, int cmd, int direction)
 		sai_start(dai, direction);
 		break;
 	case COMP_TRIGGER_STOP:
-		sai_stop(dai, direction);
-		break;
```
Comment (lgirdwood) [inline src/drivers/imx/sai.c:346]: This could be a separate PR and merged, no issues here.

File: src/ipc/ipc3/handler.c [hunk 4]
```diff
@@ -430,10 +430,10 @@ static int ipc_stream_trigger(uint32_t header)
 		cmd = COMP_TRIGGER_STOP;
 		break;
 	case SOF_IPC_STREAM_TRIG_PAUSE:
-		cmd = COMP_TRIGGER_PAUSE;
+		cmd = COMP_TRIGGER_STOP;
 		break;
 	case SOF_IPC_STREAM_TRIG_RELEASE:
-		cmd = COMP_TRIGGER_PRE_RELEASE;
+		cmd = COMP_TRIGGER_PRE_START;
```
Comment (lgirdwood) [inline src/ipc/ipc3/handler.c:436]: any issues with this ?

File: src/audio/pipeline/pipeline-schedule.c [hunk 5]
```diff
@@ -79,7 +79,7 @@ static enum task_state pipeline_task_cmd(struct pipeline *p,
 			p->status = COMP_STATE_PAUSED;
 		}
 
-		if (err == PPL_STATUS_PATH_STOP) {
+		if (err == PPL_STATUS_PATH_STOP && !p->trigger.aborted) {
```
Comment (lgirdwood) [inline src/audio/pipeline/pipeline-schedule.c:82]: Can we have a comment here on why we check both err and aborted.

**Review by lgirdwood — COMMENTED** [review]
Given there are still some issues as described above, can we split this PR into smaller chunks that can upstream until we solve 1 & 2 above.

File: src/ipc/ipc3/handler.c (hunk 4 above)
Comment (lyakh) [reply src/ipc/ipc3/handler.c:436]: We want to remove PAUSE and RELEASE, so both RELEASE and PRE_RELEASE should go

File: src/audio/pipeline/pipeline-schedule.c (hunk 5 above)
Comment (lyakh) [reply src/audio/pipeline/pipeline-schedule.c:82]: @lgirdwood thanks for checking this. The first 6 commits in this draft are actually from #5174, they'll be dropped here once that one is merged. As for the rest - this is still very much WiP. E.g. this specific code path will probably go when PAUSE and STOP are merged, or at least it will be heavily modified.

**Comment by paulstelian97** [comment]
My two cents: How does this affect pausing decoding a compressed stream? Would a stop/start work correctly (in the sense that stop won't discard the temporary state of the decoder)? And actual stopping would also do a reset which does discard said temporary state or something?

**Comment by lgirdwood** [comment]
@plbossart @ranj063 @kv2019i @lyakh the general plan is a large simplification and code reduction around our pipeline and component logic. The IPC 3/4 host interface will not change.  PAUSE == STOP and RELEASE == START.

Today all streams are stopped and paused at the next pipeline tick after the STOP IPC. This mean that the pipeline stops the DAI presentation at the next DAI DMA block boundary (i.e. its at clean DMA/cache boundaries). This is mostly the same flow as pause today (except pause has a lot more extra state machine logic). There are some pause use case specific corner cases too that need to be simplified or updated as part of this work.

The HDA state machine around stopping wont change and neither should any host IPC flows. We may need a little abstraction there to preserve any host flows but this will be a lot smaller than the code being removed..

**Comment by lgirdwood** [comment]
@paulstelian97 start/stop should not reset the processing state machine. reset only should do this, we have a feature request for aligning this that I will self assign for 2.1 or 2.2

**Comment by plbossart** [comment]
> @paulstelian97 start/stop should not reset the processing state machine. reset only should do this, we have a feature request for aligning this that I will self assign for 2.1 or 2.2

The problem is that 'reset' doesn't map to any ALSA definitions, and we're still missing the nuance between pause, stop (in the drop sense) and drain.

I am all for simplifications but if we don't have a clear view of the end goal we'll redo the work several times.

**Comment by plbossart** [comment]
> > @paulstelian97 start/stop should not reset the processing state machine. reset only should do this, we have a feature request for aligning this that I will self assign for 2.1 or 2.2
> 
> The problem is that 'reset' doesn't map to any ALSA definitions, and we're still missing the nuance between pause, stop (in the drop sense) and drain.
> 
> I am all for simplifications but if we don't have a clear view of the end goal we'll redo the work several times.

partial drain is btw a very important feature for compressed data, it's the basis for gapless playback.

**Comment by lgirdwood** [comment]
> > > @paulstelian97 start/stop should not reset the processing state machine. reset only should do this, we have a feature request for aligning this that I will self assign for 2.1 or 2.2
> > 
> > 
> > The problem is that 'reset' doesn't map to any ALSA definitions, and we're still missing the nuance between pause, stop (in the drop sense) and drain.

Yeah, ALSA PCM ops dont have a reset, but I see it as

1) Alsa pause -> IPC STOP

2) Alsa stop -> IPC STOP then IPC RESET.

> > I am all for simplifications but if we don't have a clear view of the end goal we'll redo the work several times.
> 
> partial drain is btw a very important feature for compressed data, it's the basis for gapless playback.

drain is a new case, and it should be added after the simplification work so we dont have to deal with pause and drain state machine transitions.

**Comment by paulstelian97** [comment]
Sounds good, thanks. Drain is probably good as a distinct state anyway.

**Comment by sys-pt1s** [comment]
Can one of the admins verify this patch?

File: src/ipc/Kconfig [hunk 6]
```diff
@@ -23,4 +23,12 @@ config IPC_MAJOR_4
 
 endchoice
 
+config STOP_FOR_PAUSE
+	bool
+	default BAYTRAIL || CHERRYTRAIL || HASWELL || BROADWELL
```
Comment (lgirdwood) [inline src/ipc/Kconfig:28]: Can we make this default for all atm to test the PR. It's probably going to be easier to test/fix on TGL than older HW.

**Comment by lyakh** [comment]
I tested this PR locally on a MinnowBoard BYT platform and with it the system locks up hard immediately when "release" is pressed. Debugging

**Comment by lyakh** [comment]
In this version the first patch completely disables PAUSE and RELEASE on legacy ADSPs, the second patch actually replaces the former with the latter, and I let out the HDA patch this time to remind myself what failures it was fixing.
Debugging BYT failures with the previous version resulted in a discovery of a hard system lock up during `dma_start()` for the host component, when trying to resume after the pause. The major difference with cAVS platforms is, that legacy platforms use DMA scheduling. It should be verified whether this PR causes problems on other DMA-domain platforms @dbaluta

**Comment by lyakh** [comment]
Taking the https://sof-ci.01.org/sofpr/PR5171/build745/devicetest/?model=ADLP_RVP_NOCODEC_ZEPHYR&testcase=check-pause-resume-playback-10 CI failure as an example. This seems to be the same problem as the one on legacy platforms, which is why the respective commit in this series is required. Let me try with the HDA patch now.

File: src/drivers/intel/hda/hda-dma.c [hunk 7]
```diff
@@ -685,6 +683,16 @@ static int hda_dma_stop_common(struct dma_chan_data *channel)
 	return 0;
 }
 
+static int hda_dma_host_stop(struct dma_chan_data *channel)
+{
+	hda_dma_host_stop_delayed(channel);
+	dma_chan_reg_update_bits(channel, DGCS, DGCS_FIFORDY | DGCS_GEN, 0);
```
[the remark this answers is not in the document]
Comment (lyakh) [reply src/drivers/intel/hda/hda-dma.c:689]: @plbossart this is the same flow as in `hda_dma_stop_common()` but sure, I'd welcome a review from @ranj063 too!

[the remark this answers is not in the document]
Comment (ranj063) [reply src/drivers/intel/hda/hda-dma.c:689]: @lyakh we cannot add a stop callback for the hsot DMA because it will end up stopping the DMA during the pipeline trigger STOP.

File: src/ipc/ipc3/handler.c [hunk 8]
```diff
@@ -515,10 +515,14 @@ static int ipc_glb_stream_message(uint32_t header)
 		return ipc_stream_pcm_params(header);
 	case SOF_IPC_STREAM_PCM_FREE:
 		return ipc_stream_pcm_free(header);
-	case SOF_IPC_STREAM_TRIG_START:
-	case SOF_IPC_STREAM_TRIG_STOP:
 	case SOF_IPC_STREAM_TRIG_PAUSE:
 	case SOF_IPC_STREAM_TRIG_RELEASE:
+#if CONFIG_BAYTRAIL || CONFIG_CHERRYTRAIL || CONFIG_HASWELL || CONFIG_BROADWELL
```
Comment (lgirdwood) [inline src/ipc/ipc3/handler.c:528]: The aim here is to remove this code for all platforms as PUASE is upper layer feature done by pipewire, pulseaudio and CRAS. i.e. the PR should be removing code for PAUSE/RELEASE.

File: src/drivers/intel/hda/hda-dma.c [hunk 9]
```diff
@@ -223,8 +223,6 @@ struct hda_chan_data {
 #endif
 };
 
-static int hda_dma_stop_common(struct dma_chan_data *channel);
```
Comment (lgirdwood) [inline src/drivers/intel/hda/hda-dma.c:226]: Not sure why this patch is needed if we are removing PAUSE.

**Review by lgirdwood — CHANGES_REQUESTED** [review]
This PR should be removing the logic involved with PAUSE and RELEASE.

File: src/drivers/intel/hda/hda-dma.c (hunk 9 above)
Comment (lyakh) [reply src/drivers/intel/hda/hda-dma.c:226]: @lgirdwood I think to completely remove PAUSE support we have to do that in the driver (see `hda_link_dma_trigger()`:
https://github.com/thesofproject/linux/blob/ad489ae9054724c9b61e020fa8fefd84381a9490/sound/soc/sof/intel/hda-dai.c#L288-L289
). As long as the driver treats PAUSE and STOP differently, don't think the firmware can handle them equally.

Comment (lgirdwood) [reply src/drivers/intel/hda/hda-dma.c:226]: yes, I expect this to reduce code in core and drivers too.

Comment (lyakh) [reply src/drivers/intel/hda/hda-dma.c:226]: @lgirdwood sorry, if we **remove** PAUSE that's a complete different game. This PR **keeps** PAUSE and implements it with STOP. If we **remove** PAUSE that would mean, that we return an error in response, wouldn't it?

If we remove PAUSE, we should make this somehow that the CI diesn't fail, right? So, we should first remove pause-release tests from the CI, then modify the driver to cleanly return an error. Then add a test to CI to check that PAUSE and RELEASE return that error. Then remove support for it in the firmware

Comment (lgirdwood) [reply src/drivers/intel/hda/hda-dma.c:226]: > @lgirdwood sorry, if we **remove** PAUSE that's a complete different game. This PR **keeps** PAUSE and implements it with STOP. If we **remove** PAUSE that would mean, that we return an error in response, wouldn't it?

@lyakh we need to remove pause. If driver sends pause IPC then we treat as STOP and if we cannot do this (e.g. for the HDA DMA W/A then we return error). @ranj063 can help check the HDA DMA flow.

Comment (plbossart) [reply src/drivers/intel/hda/hda-dma.c:226]: @lgirdwood @lyakh The HDAudio flow is run -> pause -> reset. There is no stop. It's going to be very very hard to map the new concepts on hardware that assumes different steps.

Comment (lgirdwood) [reply src/drivers/intel/hda/hda-dma.c:226]: @plbossart this flow is fine and will remain. My main worry is that we had flows with both pause and stop combined to form a correct IP programming sequence.

Comment (lyakh) [reply src/drivers/intel/hda/hda-dma.c:226]: > Not sure why this patch is needed if we are removing PAUSE.

@lgirdwood yes, we remove the PAUSE trigger, i.e. `COMP_TRIGGER_PAUSE` - as long as `CONFIG_STOP_FOR_PAUSE` is enabled, which is the default.
But when the driver sends a PAUSE IPC upon a "pause" request from the user, we use `COMP_TRIGGER_STOP` in the SOF. Then, when a RELEASE IPC comes from the driver, we convert it to `COMP_TRIGGER_PRE_START`. But this alone doesn't work with HDA, because when a real STOP arrives, it doesn't yet stop the HDA DMA, and it is only stopped when `PCM_FREE` arrives. Since with PAUSE we don't have `PCM_FREE` we need this patch for HDA DMA.

**Review by lgirdwood — COMMENTED** [review]
@lyakh lets figure out the driver changes so that the FW changes are 99% code removal.

File: src/drivers/intel/hda/hda-dma.c (hunk 9 above)
Comment (plbossart) [reply src/drivers/intel/hda/hda-dma.c:226]: @lgirdwood we've discussed this some 2 years ago. Windows requires pause to be supported, so I really don't see where you are going with removing Pause. It's a non-starter.

Comment (plbossart) [reply src/drivers/intel/hda/hda-dma.c:226]: https://github.com/thesofproject/linux/issues/1339

Comment (lgirdwood) [reply src/drivers/intel/hda/hda-dma.c:226]: @plbossart Windows uses PAUSE IPC to STOP pipelines. It has no STOP named IPC.

e.g. the pipeline states.
```c
/**< Pipeline State */
enum ipc4_pipeline_state {
	/**< Invalid value */
	SOF_IPC4_PIPELINE_STATE_INVALID = 0,
	/**< Created but initialization incomplete */
	SOF_IPC4_PIPELINE_STATE_UNINITIALIZED = 1,
	/**< Resets pipeline */
	SOF_IPC4_PIPELINE_STATE_RESET = 2,
	/**< Pauses pipeline */
	SOF_IPC4_PIPELINE_STATE_PAUSED = 3,
	/**< Starts pipeline */
	SOF_IPC4_PIPELINE_STATE_RUNNING = 4,
	/**< Marks pipeline as expecting End Of Stream */
	SOF_IPC4_PIPELINE_STATE_EOS = 5,
	/**< Stopped on error */
	SOF_IPC4_PIPELINE_STATE_ERROR_STOP, <<<<<< this one is used to report a FW error to host.
	/**< Saved to the host memory */
	SOF_IPC4_PIPELINE_STATE_SAVED
};
```

and the SOF driver trigger logic 

```c
static int sof_ipc4_pcm_trigger(struct snd_soc_component *component,
				struct snd_pcm_substream *substream, int cmd)
{
	int state;

	/* determine the pipeline state */
	switch (cmd) {
	case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
		state = SOF_IPC4_PIPE_PAUSED;
		break;
	case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
	case SNDRV_PCM_TRIGGER_RESUME:
	case SNDRV_PCM_TRIGGER_START:
		state = SOF_IPC4_PIPE_RUNNING;
		break;
	case SNDRV_PCM_TRIGGER_SUSPEND:
	case SNDRV_PCM_TRIGGER_STOP:
		state = SOF_IPC4_PIPE_PAUSED;
		break;
	default:
		dev_err(component->dev, "%s: unhandled trigger cmd %d\n", __func__, cmd);
		return -EINVAL;
	}

	/* set the pipeline state */
	return sof_ipc4_trigger_pipelines(component, substream, state);
}
```

Comment (plbossart) [reply src/drivers/intel/hda/hda-dma.c:226]: It is a converged firmware requirement to support pause. You cannot remove this code now, or you will have to re-add it later.

It's also very useful at the kernel level to deal with all kinds of issues with suspend-resume.

I don't see any rationale for removing SNDRV_PCM_INFO_PAUSE. It has shown multiple issues in our state machines and forced us multiple times to do the right thing for suspend-resume.

**Comment by lyakh** [comment]
> @lyakh lets figure out the driver changes so that the FW changes are 99% code removal.

@lgirdwood so we should block PAUSE / RELEASE at the driver level? But since IPCs exist in both IPC3 and IPC4 I think the firmware still should be able to handle them, e.g. when loaded with an older kernel. So we can already just modify the firmware to return errors to those IPCs, can we not?

File: src/drivers/intel/hda/hda-dma.c (hunk 9 above)
Comment (plbossart) [reply src/drivers/intel/hda/hda-dma.c:226]: Sorry @lgirdwood the existing kernel sequences are part of the ABI. We don't control when kernel patches are integrated, which makes it impossible for the firmware to unilaterally decide when the changes are implemented. And since in most cases the firmware is updated much faster than the kernel you will most likely have to keep the sequences as is for quite a while.

**Comment by kv2019i** [comment]
No activity in a long time, converting as draft.

**Comment by kv2019i** [comment]
Stale PR,closing
